src/java.base/share/classes/java/util/spi/ToolProvider.java

```diff
@@ -50,6 +50,14 @@
  * arguments that could be provided to the tool when invoking the tool
  * from the command line.
  *
+ * @apiNote
+ * It is recommended that tools implementing this interface are either
+ * reusable and reentrant, or should clearly document any limitations and
+ * restrictions. In this context, reusable means that any one instance of
+ * a tool may be the target of multiple {@code run} method invocations,
+ * and reentrant means that multiple invocations of {@code run} may occur
+ * concurrently.
```

Comment:
Hello @sormuras,

> reentrant means that multiple invocations of {@code run} may occur
> concurrently.

My understanding of re-entrant was that the same method could be re-invoked on the same thread while the current method execution is in progress (a recursion). Whereas "multiple invocations may occur concurrently" sounds more like multiple threads invoking this concurrently (i.e. thread-safety). Which of these 2 characteristics are we recommending here?

Reply:
According to https://wikidiff.com/reentrant/reusable "reentrant" includes spans both characteristics:

> (programming) That may be executed more than once at a time either by different threads, or because of recursion.

Reply:
I agree with @jaikiran that reentrant is much more common in the sense of "on the same thread". Plus that the JDK itself uses "ReentrantLock" in this meaning. The JDK uses either "thread-safe" or "synchronized" for the "multiple threads" meaning.

Actually, being thread-safe implies being reusable, so I'd just phrase it as:

> It is recommended that tools implementing this interface are either
> thread-safe, or clearly document any limitations and restrictions.

Also: should this be an `@implNote`, rather than an `@apiNote`, since it's about "tools implementing this interface"?